### ISSUE TITLE
[CI] `linux://python/ray/tests:kuberay/test_autoscaling_e2e` is failing/flaky on master.

### ISSUE THREAD

**Issue by cadedaniel** [ray-project/ray#35471]
- 4202bc2a38059d2bfba53e8639f6842eb3480bda FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3966#01882b57-e6f5-4f77-9613-5aa723f27a2c)
- 987a1e35fd1f681d4431d0b279507bc5d3657748 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3965#01882b45-8e55-4293-8d83-a3dc952ef0f0)
- 9ac0d44769fc032d504818f9c77931f19eeccb52 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3964#01882b1e-6681-4cff-b209-1c31c2ab9856)
- 5ec3a36c8266229c3888494fda483083a175e6db FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3963#01882b12-89a2-447a-962a-6d96f5469b7f)
- 45067ae21835a78666f0da1143ec4469c6265667 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3962#01882af3-c299-4cf0-a35c-9ac2ea7f06c4)
- ae5e1c7acca1073f6e16d7428e46f46e2535d2e8 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3961#01882ac6-3bcb-464c-88bb-117e420c294b)
- 1bd30c32974f0827c3a42bf328c95c719f5e74b3 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3960#01882a4d-84e6-4464-91b8-a258743caa8f)
- cde7109a244ce40ed15a7259669b55301a6ace74 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3959#01882a0e-dea8-4c91-b885-60a50426a9a2)
- 119e12bb5d1b009f791d7cc4d52bb790196640c7 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3958#018828d7-edcf-4282-8935-aa66ea02fee2)
- 9cd509723494eb861c77a989cfa880ec384d31c6 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3957#018827a8-54a4-4dee-82c8-3a246d6fe52d)
- a9f333a9068b37951b2c2d827f1e9f0c39123581 FAILED [Buildkite :kubernetes: operator](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3956#01882795-fd04-4b7d-8402-c3008ea42c07)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-linux://python/ray/tests:kuberay/test_autoscaling_e2e-END
....

**Comment by ArturNiederfahrenhorst** [comment]
@cadedaniel @kevin85421 
This is also broken on the release branch.
Can you make sure that a fix lands soonish and ping me for a pick?

**Comment by kevin85421** [comment]
Hi @ArturNiederfahrenhorst, I am starting to work on this issue. I will try to figure out the issue asap.

**Comment by kevin85421** [comment]
cc @architkulkarni

**Comment by architkulkarni** [comment]
<img width="1701" alt="Screenshot 2023-05-18 at 4 08 47 PM" src="https://github.com/ray-project/ray/assets/5459654/e187bca2-7e82-4f96-ad15-38bb1f274af7">
The commit at which it started failing is almost certainly unrelated: https://github.com/ray-project/ray/pull/35401

  Here's the traceback of the failure:
  ```
          # 4. Confirm that the GPU actor is up and that Ray believes
        # the node the actor is on has a GPU.
        logger.info("Confirming GPU actor placement.")
        with ray_client_port_forward(
            head_service=HEAD_SERVICE, ray_namespace="gpu-test"
        ):
>           out = gpu_actor_validation.main()
/root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/python/ray/tests/kuberay/test_autoscaling_e2e.runfiles/com_github_ray_project_ray/python/ray/tests/kuberay/test_autoscaling_e2e.py:350:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/ray/python/ray/tests/kuberay/scripts/gpu_actor_validation.py:7: in main
    actor_response = ray.get(gpu_actor.where_am_i.remote())
/ray/python/ray/_private/auto_init_hook.py:18: in auto_init_wrapper
    return fn(*args, **kwargs)
/ray/python/ray/_private/auto_init_hook.py:18: in auto_init_wrapper
    return fn(*args, **kwargs)
/ray/python/ray/_private/client_mode_hook.py:102: in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
/ray/python/ray/util/client/api.py:42: in get
    return self.worker.get(vals, timeout=timeout)
/ray/python/ray/util/client/worker.py:434: in get
    res = self._get(to_get, op_timeout)
/ray/python/ray/util/client/worker.py:455: in _get
    for chunk in resp:
/ray/python/ray/util/client/worker.py:324: in _get_object_iterator
    for chunk in self.server.GetObject(req, *args, **kwargs):
/opt/miniconda/lib/python3.7/site-packages/grpc/_channel.py:426: in __next__
    return self._next()
/opt/miniconda/lib/python3.7/site-packages/grpc/_channel.py:817: in _next
    _common.wait(self._state.condition.wait, _response_ready)
/opt/miniconda/lib/python3.7/site-packages/grpc/_common.py:150: in wait
    _wait_once(wait_fn, MAXIMUM_WAIT_TIMEOUT, spin_cb)
/opt/miniconda/lib/python3.7/site-packages/grpc/_common.py:112: in _wait_once
    wait_fn(timeout=timeout)
/opt/miniconda/lib/python3.7/threading.py:300: in wait
    gotit = waiter.acquire(True, timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
signum = 15
frame = <frame at 0x560b27ea4ef0, file '/opt/miniconda/lib/python3.7/threading.py', line 310, code wait>
    def sigterm_handler(signum, frame):
>       sys.exit(signum)
E       SystemExit: 15
/ray/python/ray/_private/worker.py:1726: SystemExit
```
   (For anyone else looking at buildkite logs directly, note that subsequent failures on retries 2/3 and 3/3 are a red herring, due to bad or missing cleanup logic.)
   
   Not clear yet why this timeout suddenly started happening. If we can't figure it out, one option is to rewrite this portion to not use Ray Client.

**Comment by can-anyscale** [comment]
Since this is a release-blocker issue, please **close it only after the cherry pick fix is merged into 2.5 release branch**. 

Please add @ArturNiederfahrenhorst as one of the reviewer of the fix as well for tracking purpose. Thankks!

**Comment by architkulkarni** [comment]
<img width="1676" alt="Screenshot 2023-05-22 at 11 09 54 AM" src="https://github.com/ray-project/ray/assets/5459654/e2beaa13-b954-471e-8880-9ac7f29c2aff">
The problem appears to have fixed itself on the master branch and on the release branch (e.g. https://buildkite.com/ray-project/oss-ci-build-branch/builds/4033#01884464-fd93-418c-88c0-d7823d50d050).  The "breaking commit" and the "un-breaking commit" are almost certainly both unrelated.  My best guess is some unpinned downstream dependency (of `grpcio`, maybe) had a buggy update which got hotfixed.

In case this problem comes up again, here's our notes from debugging with @kevin85421  (summarized by @kevin85421 ):

> More details: We utilize ray health-check with the Kubernetes FQDN (fully qualified domain name) as the RAY_ADDRESS. However, the health-check reports the following DNS error:
(base) ray@raycluster-autoscaler-worker-fake-gpu-group-6pt7z:~$ ray health-check --address $FQ_RAY_IP:6379
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/gcs_utils.py", line 123, in check_health
    resp = stub.CheckAlive(req, timeout=timeout)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "DNS resolution failed for raycluster-autoscaler-head-svc.default.svc.cluster.local:6379: C-ares status is not ARES_SUCCESS qtype=AAAA name=raycluster-autoscaler-head-svc.default.svc.cluster.local is_balancer=0: Could not contact DNS servers"
        debug_error_string = "UNKNOWN:DNS resolution failed for raycluster-autoscaler-head-svc.default.svc.cluster.local:6379: C-ares status is not ARES_SUCCESS qtype=AAAA name=raycluster-autoscaler-head-svc.default.svc.cluster.local is_balancer=0: Could not contact DNS servers {created_time:"2023-05-19T15:41:25.366279004-07:00", grpc_status:14}"
>
> The issue appears to be related to a dependency on gRPC, as we conducted the following experiments to investigate further:
FQ_RAY_IP: The head Pod’s DNS in FQDN
RAY_IP: The head Pod’s DNS (only have the Kubernetes Service name)
curl $RAY_IP:8265 => Succeed
curl $FQ_RAY_IP:8265 => Succeed
ray health-check --address $RAY_IP:6379 => Succeed
ray health-check --address $FQ_RAY_IP:6379 => Fail
ray job submit --address http://$RAY_IP:8265 … => Succeed
ray job submit --address http://$FQ_RAY_IP:8265 … => Succeed
It has been observed that the Kubernetes cluster’s DNS server, CoreDNS, is capable of resolving both $FQ_RAY_IP and $RAY_IP. However, ray health-check cannot resolve FQDN with C-ares. Has there been any recent update to the gRPC version?

Some links which may prove useful for debugging in the future, in case this happens again (since the fix has happened somewhere unknown outside the Ray repo, we won't be able to reproduce the issue on demand any more):
https://github.com/microsoft/azure-container-apps/issues/45
https://github.com/coredns/coredns/issues/2389 
https://github.com/grpc/grpc/issues/28870
https://github.com/telepresenceio/telepresence/issues/1871#issuecomment-1441959400 

The workarounds suggested are to set `GRPC_DNS_RESOLVER=native`, or call a target URL without a specified scheme (e.g. `http://`). But what I got from these is that these are only workarounds, and don't explain why this suddenly broke or how it got fixed.

**Comment by architkulkarni** [comment]
> Since this is a release-blocker issue, please **close it only after the cherry pick fix is merged into 2.5 release branch**.
> 
> Please add @ArturNiederfahrenhorst as one of the reviewer of the fix as well for tracking purpose. Thankks!

@can-anyscale as this is a special case (fixed itself due to some change outside the Ray repo), there is no cherry-pick fix needed for the release.

**Comment by can-anyscale** [comment]
Best problem to have, thanks @architkulkarni !
